OpenZFS 8585 - improve batching done in zil_commit()

prakashsurya opened the pull request:
Authored by: Prakash Surya <prakash.surya@delphix.com>
Reviewed by: Brad Lewis <brad.lewis@delphix.com>
Reviewed by: Matt Ahrens <mahrens@delphix.com>
Reviewed by: George Wilson <george.wilson@delphix.com>
Ported-by: Prakash Surya <prakash.surya@delphix.com>

Problem
=======

The current implementation of zil_commit() can introduce significant
latency, beyond what is inherent due to the latency of the underlying
storage. The additional latency comes from two main problems:

 1. When there's outstanding ZIL blocks being written (i.e. there's
    already a "writer thread" in progress), then any new calls to
    zil_commit() will block waiting for the currently oustanding ZIL
    blocks to complete. The blocks written for each "writer thread" is
    coined a "batch", and there can only ever be a single "batch" being
    written at a time. When a batch is being written, any new ZIL
    transactions will have to wait for the next batch to be written,
    which won't occur until the current batch finishes.

    As a result, the underlying storage may not be used as efficiently
    as possible. While "new" threads enter zil_commit() and are blocked
    waiting for the next batch, it's possible that the underlying
    storage isn't fully utilized by the current batch of ZIL blocks. In
    that case, it'd be better to allow these new threads to generate
    (and issue) a new ZIL block, such that it could be serviced by the
    underlying storage concurrently with the other ZIL blocks that are
    being serviced.

 2. Any call to zil_commit() must wait for all ZIL blocks in its "batch"
    to complete, prior to zil_commit() returning. The size of any given
    batch is proportional to the number of ZIL transaction in the queue
    at the time that the batch starts processing the queue; which
    doesn't occur until the previous batch completes. Thus, if there's a
    lot of transactions in the queue, the batch could be composed of
    many ZIL blocks, and each call to zil_commit() will have to wait for
    all of these writes to complete (even if the thread calling
    zil_commit() only cared about one of the transactions in the batch).

To further complicate the situation, these two issues result in the
following side effect:

 3. If a given batch takes longer to complete than normal, this results
    in larger batch sizes, which then take longer to complete and
    further drive up the latency of zil_commit(). This can occur for a
    number of reasons, including (but not limited to): transient changes
    in the workload, and storage latency irregularites.

Solution
========

The solution attempted by this change has the following goals:

 1. no on-disk changes; maintain current on-disk format.
 2. modify the "batch size" to be equal to the "ZIL block size".
 3. allow new batches to be generated and issued to disk, while there's
    already batches being serviced by the disk.
 4. allow zil_commit() to wait for as few ZIL blocks as possible.
 5. use as few ZIL blocks as possible, for the same amount of ZIL
    transactions, without introducing significant latency to any
    individual ZIL transaction. i.e. use fewer, but larger, ZIL blocks.

In theory, with these goals met, the new allgorithm will allow the
following improvements:

 1. new ZIL blocks can be generated and issued, while there's already
    oustanding ZIL blocks being serviced by the storage.
 2. the latency of zil_commit() should be proportional to the underlying
    storage latency, rather than the incoming synchronous workload.

Porting Notes
=============

Due to the changes made in commit 119a394ab0, the lifetime of an itx
structure differs than in OpenZFS. Specifically, the itx structure is
kept around until the data associated with the itx is considered to be
safe on disk; this is so that the itx's callback can be called after the
data is committed to stable storage. Since OpenZFS doesn't have this itx
callback mechanism, it's able to destroy the itx structure immediately
after the itx is committed to an lwb (before the lwb is written to
disk).

To support this difference, and to ensure the itx's callbacks can still
be called after the itx's data is on disk, a few changes had to be made:

  * A list of itxs was added to the lwb structure. This list contains
    all of the itxs that have been committed to the lwb, such that the
    callbacks for these itxs can be called from zil_lwb_flush_vdevs_done(),
    after the data for the itxs is committed to disk.

  * A list of itxs was added on the stack of the zil_process_commit_list()
    function; the "nolwb_itxs" list. In some circumstances, an itx may
    not be committed to an lwb (e.g. if allocating the "next" ZIL block
    on disk fails), so this list is used to keep track of which itxs
    fall into this state, such that their callbacks can be called after
    the ZIL's writer pipeline is "stalled".

  * The logic to actually call the itx's callback was moved into the
    zil_itx_destroy() function. Since all consumers of zil_itx_destroy()
    were effectively performing the same logic (i.e. if callback is
    non-null, call the callback), it seemed like useful code cleanup to
    consolidate this logic into a single function.

Additionally, the existing Linux tracepoint infrastructure dealing with
the ZIL's probes and structures had to be updated to reflect these code
changes. Specifically:

  * The "zil__cw1" and "zil__cw2" probes were removed, so they had to be
    removed from "trace_zil.h" as well.

  * Some of the zilog structure's fields were removed, which affected
    the tracepoint definitions of the structure.

  * New tracepoints had to be added for the following 3 new probes:
      * zil__process__commit__itx
      * zil__process__normal__itx
      * zil__commit__io__error

OpenZFS-issue: https://www.illumos.org/issues/8585
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/5d95a3a

prakashsurya commented:
@sempervictus this is a second attempt at porting what I previously had in #6337, and attempts to address the same performance issues as #6133 .. if you can, it'd be great if you could try this out in your test environment.

I've fixed the my prior issues hit with the zfstest suite, and was able to successfully run this version through the test suite in a local VM. Additionally, I believe the build issues that you previously hit w.r.t. the new tracepoints have been fixed (tested and verified they worked in my VM).

The OpenZFS PR contains links to data on the performance testing that I previously did for this change (on illumos); see the "Performance Testing" section of that PR's description: https://github.com/openzfs/openzfs/pull/447

sempervictus commented:
@prakashsurya: adding to the test queue, thanks for digging in so quickly.

prakashsurya commented:
@behlendorf is the `nopwrite/nopwrite_volume` test reliable? that appears to have failed, so I'm curious if that's a known failure mode in the existing code, or if it's likely a result of my changes (and I should dig into the root cause). here's what the logs shows:
```
Test: /usr/share/zfs/zfs-tests/tests/functional/nopwrite/nopwrite_volume (run as root) [00:14] [FAIL]
05:35:21.18 ASSERTION: nopwrite works on volumes
05:35:21.33 SUCCESS: zfs set compress=on testpool/testvol.nopwrite
05:35:21.42 SUCCESS: zfs set checksum=sha256 testpool/testvol.nopwrite
05:35:31.03 SUCCESS: zfs clone testpool/testvol.nopwrite@a testpool/clone
05:35:31.33 SUCCESS: zfs set compress=on testpool/clone
05:35:31.43 SUCCESS: zfs set checksum=sha256 testpool/clone
05:35:33.59 SUCCESS: zpool sync testpool
05:35:33.76 NOTE: Comparing 33886720 and 377856 given 2% (calculated: 1.11%)
05:35:33.78 NOTE: Comparing 33907712 and 0 given 2% (calculated: 0%)
05:35:33.79 NOTE: Comparing 33886720 and 377856 given 2% (calculated: 1.11%)
05:35:33.81 NOTE: Comparing 33907712 and 33886720 given 98% (calculated: 99.93%)
05:35:33.83 NOTE: Comparing 33886720 and 33886720 given 98% (calculated: 100.00%)
05:35:35.12 NOTE: Comparing 377856 and 356864 given 98% (calculated: 94.44%)
05:35:35.13 ERROR: verify_nopwrite testpool/testvol.nopwrite testpool/testvol.nopwrite@a testpool/clone exited 1
```

behlendorf commented:
@prakashsurya it hasn't been 100% reliable and we've opened #5510 to track the known failure.  We see a failure maybe 1/50 runs on the 32-bit tester but haven't had a chance completely root cause and fix the issue.  Let me run it through again and see if it's reproducible, below is the link to the previous test failures so we don't loose track of it.

http://build.zfsonlinux.org/builders/Ubuntu%2014.04%20i686%20%28TEST%29/builds/3556

sempervictus commented:
A bunch of zloop later, still haven't blown it up. Going to push to some metal today, see how it fares on real world ops.

sempervictus commented:
This is promising - two personal production systems now running with this PR, send/recv working, zvols being used for KVM instances. Contended writes are a lot better - double digit reduction in iowait with multiple VMS running FIO. Thank you @prakashsurya

prakashsurya commented:
appreciate the feedback; thanks @sempervictus.

behlendorf commented:
@prakashsurya it looks like there's still a failure on the 32-bit testers which needs to be run down.  I [ran the tests](http://build.zfsonlinux.org/builders/Ubuntu%2014.04%20i686%20%28TEST%29/builds/3601) a few times to make sure and it's reproducible.  The test case `zpool_clear_001_pos` appears to cause the issue and I was able to pull the following stack out of the ZTS logs.  Unfortunately, the exact ASSERT we hit was lost.  

```
z_vdev_file     D 00000000     0 15249      2  
Call Trace:
 [<c16633c3>] schedule+0x23/0x60
 [<d5606485>] spl_panic+0xc5/0xe0 [spl]
 [<d5ddbac2>] zfs_get_data+0x1b2/0x570 [zfs] 
 [<d5dea586>] zil_commit+0x1266/0x1fd0 [zfs] 
 [<d5dd32dc>] zfs_fsync+0x6c/0xd0 [zfs]
 [<d5e061af>] zpl_fsync+0x7f/0x130 [zfs]
 [<c11a6f9f>] vfs_fsync+0x2f/0x40
 [<d560591a>] vn_fsync+0x3a/0xd0 [spl]
 [<d5d914d0>] vdev_file_io_fsync+0x30/0x50
 [<d5603997>] taskq_thread+0x267/0x5a0 [spl] 
 [<c1076fdb>] kthread+0x9b/0xb0
```

prakashsurya commented:
Thanks for the report. Briefly looking at the assertion, I don't know why that'd be occurring; I'll have to dig more into it and maybe set up an environment to reproduce it locally. It appears as though the write record passed into `zfs_get_data` has an `lr_length` of zero, but I don't immediately know how/why that would occur (nor why it'd only occur on the 32-bit system). This needs more investigation.

sempervictus commented:
@prakashsurya: saw similar error in a long and laggy send/recv. Not sure if its the patch or a very slow recv host though - hung task didn't crash hosts. The original 64b p4 in there is not loving it. Shows how far chips have come in a decade I guess.

prakashsurya commented:
While running some tests of this patch in a VM, I managed trip the following assertion:
```
[ 1806.416279] VERIFY3(itx->itx_wr_state == WR_INDIRECT) failed (4294967295 == 0)
[ 1806.416315] PANIC at zil.c:1506:zil_lwb_commit()
[ 1806.416333] Showing stack for process 28365
[ 1806.416336] CPU: 5 PID: 28365 Comm: fio Tainted: P           OE   4.4.0-93-generic #116-Ubuntu
[ 1806.416337] Hardware name: VMware, Inc. VMware Virtual Platform/440BX Desktop Reference Platform, BIOS 6.00 09/21/2015
[ 1806.416338]  0000000000000286 16a1c125c5c4eec7 ffff887cd21678c0 ffffffff813f9f83
[ 1806.416341]  ffffffffc136d454 00000000000005e2 ffff887cd21678d0 ffffffffc02d6fd2
[ 1806.416342]  ffff887cd2167a58 ffffffffc02d70a5 ffff887b665d75d0 ffff887c00000030
[ 1806.416344] Call Trace:
[ 1806.416350]  [<ffffffff813f9f83>] dump_stack+0x63/0x90
[ 1806.416362]  [<ffffffffc02d6fd2>] spl_dumpstack+0x42/0x50 [spl]
[ 1806.416366]  [<ffffffffc02d70a5>] spl_panic+0xc5/0x100 [spl]
[ 1806.416428]  [<ffffffffc127ab1e>] ? zio_rewrite+0x6e/0x70 [zfs]
[ 1806.416469]  [<ffffffffc126bd90>] ? zil_lwb_write_open+0x520/0x520 [zfs]
[ 1806.416509]  [<ffffffffc126b96f>] ? zil_lwb_write_open+0xff/0x520 [zfs]
[ 1806.416551]  [<ffffffffc126bd90>] ? zil_lwb_write_open+0x520/0x520 [zfs]
[ 1806.416591]  [<ffffffffc1271daa>] zil_commit+0x14ba/0x1d90 [zfs]
[ 1806.416635]  [<ffffffffc1260d6d>] zfs_write+0xe0d/0x1020 [zfs]
[ 1806.416677]  [<ffffffffc128b391>] zpl_write_common_iovec+0x91/0x130 [zfs]
[ 1806.416717]  [<ffffffffc128b4e1>] zpl_iter_write+0xb1/0xf0 [zfs]
[ 1806.416721]  [<ffffffff8120ec0b>] new_sync_write+0x9b/0xe0
[ 1806.416723]  [<ffffffff8120ec76>] __vfs_write+0x26/0x40
[ 1806.416724]  [<ffffffff8120f5f9>] vfs_write+0xa9/0x1a0
[ 1806.416726]  [<ffffffff81210465>] SyS_pwrite64+0x95/0xb0
[ 1806.416729]  [<ffffffff81039d03>] ? fpu__restore+0xf3/0x150
[ 1806.416734]  [<ffffffff818431f2>] entry_SYSCALL_64_fastpath+0x16/0x71
```
this was running on a VM with 8 SSDs, 32 vCPUs, 512GB of RAM, and 512 FIO threads performing synchronous writes; I'm not yet sure what the root cause for this is.

sempervictus commented:
I believe I saw that on a prior revision of this PR. Couldn't repro however.

prakashsurya commented:
OK, that's good to know. I can reproduce this every time I run this workload on this system, so I hope to root cause it sometimes this week (it's a shame not having `mdb` at my disposal).

sempervictus commented:
We've had this running for a few weeks including a few production hosts backing cinder with pcie ssds for slog on spinning pools. Writes are definitely snappier and the stalls are greatly reduced under concurrent load. If there's no blocker outstanding, I think this should get merged for wider eval.

kernelOfTruth commented:
Just to understand this a bit better: this also applies with no ZIL device connected, right ?

In that case I can't wait to profit from those lower latencies and higher IOPS

Thanks a lot !

prakashsurya commented:
@kernelOfTruth The "ZIL" cannot be removed, so I think you meant "SLOG" right? If so, yes, this should improve latency even with no SLOG connected. My performance tests done on illumos/delphixos were done without an SLOG, and showed improved latencies; see [here](https://www.prakashsurya.com/post/2017-09-08-performance-testing-results-for-openzfs-447/) for more details.

@sempervictus There's two blockers which I haven't yet gotten around to fixing:

1. I hit a panic when running my performance tests on an Ubuntu VM with an all SSD pool, and 32 vCPUs. See [here](
https://github.com/zfsonlinux/zfs/pull/6566#issuecomment-327270825).

2. Brian hit a panic when running on a 32-bit Ubuntu VM. See [here](https://github.com/zfsonlinux/zfs/pull/6566#issuecomment-326106924).

IMO, both of these should be investigated and fixed prior to landing.

kernelOfTruth commented:
@prakashsurya yes, sorry I meant SLOG device, all of the time I was reading on the web about it only was called "ZIL device"

[the following seems to be one article that uses the correct term https://www.ixsystems.com/blog/o-slog-not-slog-best-configure-zfs-intent-log/ ]

sempervictus commented:
@prakashsurya: regarding the crash you experienced, i'm rather suspicious of your hypervisor. Type 2 HVs are known for doing "the wrong thing" when it comes to IO, specifically as relating to synchronous semantics/commit order/other things sane people expect to work. VBox is most guilty of this, but any type 2 is suspect. Can you repro on ESXi?
Far as the failure on 32bit goes, i'm glad to see some support for the platform, but the vast majority of real world consumers are 64bit. Maybe leave in the existing codepath defined out only for 32bit? Are newer 32bit distros seeing this as well? Compiler version does make a difference...
On a related note - I've realized that running Arch Linux (rolling distro) i'm always getting the newest version of gcc (which makes for a nightmarish experience with randstruct and other plugins affecting my kernel and dkms modules). Apaprently there's benefit over the older versions found in more mainline distributions in terms of compatibility and performance. There's a good example of this on the [kernel-hardening list](https://patchwork.kernel.org/project/kernel-hardening/list/) in discussions around PIE performance impact. It might make sense to have a builder on such a distribution which can produce packages for other distros so we can compare them against natively built versions when we have reproducers (to track down the problem to the distro or code). It would also give us "current baseline" binaries for each tested commit, which may not hurt either.

kernelOfTruth commented:
@sempervictus no worries - now I know the internals a bit better :)

Makes sense to employ the change where it matters the most - ZFS and the hardware must be beaten pretty heavily in that usage scenario - that'll (as shown) most likely also will reveal any existing, dormant issues (if applicable) 👍

prakashsurya commented:
@sempervictus For my performance testing, and the panic I linked to, I was using VMware for the hypervisor. It was the same server/hypervisor that I used for the tests that I ran on illumos.

behlendorf commented:
@prakashsurya if you get a chance can you rebase this one more time on master.  I'd like to see if the 32-bit fix from #6734, which was merged today, has any impact of the i686 test results.

prakashsurya commented:
@behlendorf done.

behlendorf commented:
That's a bit better, no ASSERT.  Those all look like known spurious test failures, which suggests the root cause for this issue may be aggravated by a low memory situation.  I'll send the build through a few more times to see if we always get the same set of test case failures.

prakashsurya commented:
To make sure I understand, you're saying the 32-bit failure was due to an unrelated bug (i.e. not in this patch) that's been fixed? (pending re-running the tests a few more times)

behlendorf commented:
Not exactly.  The 32-bug, which was recently fixed, caused the ARC to miscalculate the amount of available free memory because it wasn't excluding free highmem pages.  As a result the ARC wasn't shrinking when it needed to which put low memory stress on the rest of the system.  This definitely resulted in OOM events and possibly memory allocation failures.  How that ended up causing the ASSERT isn't clear.

However, since resolving that issue I've been unable to reproduce the ASSERT reported above in any of my 32-bit testing.  The additional buildbot runs (4 so far) also only showed known spurious unrelated ZTS failures.  And in fact we did get one [entirely clean test run](http://build.zfsonlinux.org/builders/Ubuntu%2014.04%20i686%20%28TEST%29/builds/4184).

My suspicion is that the flaw isn't 32-bit specific but was simply easier to reproduce there do to this other problem.  The ASSERT probably could happen on an x86_64 system too given the right circumstances.  But since we don't actually know the root cause that's all really just guess work.

@sempervictus @prakashsurya I'd be open to merging this for broader testing, assuming we're no longer able to reproduce either of the two above ASSERTs.  And we can verify the merging this doesn't increase the frequency of those unrelated ZTS test failures.

sempervictus commented:
@prakashsurya: trying to get a fresh patch stack set up, and having some trouble merging this. Could i ask for a rebase on master? Thank you.

prakashsurya commented:
Thanks for the heads up. If the merge isn't too difficult, I can probably do that in a few hours.

prakashsurya commented:
I pushed a new version that I think will work. I haven't tested (or even built it), so let's see if the automated tests report any issues with the rebased version.

sempervictus commented:
@prakashsurya: trying to get a fresh patch stack set up, and having some trouble merging this. Could i ask for a rebase on master? Thank you.

sempervictus commented:
@prakashsurya: so it seems to work on _most_ of the test hosts, but one consistently crashes after ~12-20 hours with [this gem](https://pasteboard.co/GSUKrjr.png). I think this is being caused by the removal of lock tracking - #6107,  in conjunction with this PR, but definitely something to watch for with both of them in play. The system is known to have an SMP quirk, it's acted out before, so this may all go away on its own, but if there's a real null deref in there, its probably best to catch it now :).

sempervictus commented:
@prakashsurya: i'm seeing ztest failures in testing VMs, but physical systems seem ok:
```
  0.89 sec in ztest_dmu_read_write
error: metaslab_free_dva(): bad DVA 6422528:14123288431444754944:10878976
/usr/bin/ztest(+0x9cbf)[0x7d26d21cbf]
/usr/lib/libpthread.so.0(+0x11da0)[0x35ae8d5bda0]
/usr/lib/libc.so.6(gsignal+0x110)[0x35ae89c68a0]
/usr/lib/libc.so.6(abort+0x1c9)[0x35ae89c7f09]
/usr/lib/libzpool.so.2(+0x3ff32)[0x35ae92f3f32]
/usr/lib/libzpool.so.2(vcmn_err+0x4b)[0x35ae92f405b]
/usr/lib/libzpool.so.2(zfs_panic_recover+0xbf)[0x35ae9377aef]
/usr/lib/libzpool.so.2(metaslab_free+0xf5)[0x35ae935ddd5]
/usr/lib/libzpool.so.2(+0x14898f)[0x35ae93fc98f]
/usr/lib/libzpool.so.2(zio_wait+0xab)[0x35ae940161b]
/usr/lib/libzpool.so.2(zio_free+0xae)[0x35ae940176e]
/usr/lib/libzpool.so.2(dsl_dataset_block_kill+0x273)[0x35ae9330d33]
/usr/lib/libzpool.so.2(+0x7a243)[0x35ae932e243]
/usr/lib/libzpool.so.2(+0x7a90e)[0x35ae932e90e]
/usr/lib/libzpool.so.2(range_tree_vacate+0x76)[0x35ae9360be6]
/usr/lib/libzpool.so.2(dnode_sync+0x289)[0x35ae932ed89]
/usr/lib/libzpool.so.2(+0x65f4a)[0x35ae9319f4a]
/usr/lib/libzpool.so.2(+0x40bc6)[0x35ae92f4bc6]
/usr/lib/libpthread.so.0(+0x708a)[0x35ae8d5108a]
/usr/lib/libc.so.6(clone+0x3f)[0x35ae8a8847f]
child died with signal 6
```
Is that coming from this PR or master?

prakashsurya commented:
Those ztest stacks don't look related to my changes in this PR, but without a root cause, we can't 100% rule this change out.

sempervictus commented:
Figured out how to get the textual stack trace from the zil_commit crashes - grsec locks out sshd since it runs as UID99, but LXC containers can still get dmesg out :). So here's what i'm seeing on the physical crashes:
```
[Fri Nov 10 14:33:45 2017] PAX: please report this to pageexec@freemail.hu
[Fri Nov 10 14:33:45 2017] BUG: unable to handle kernel NULL pointer dereference at            (nil)
[Fri Nov 10 14:33:45 2017] IP: [<ffffffff813d2cae>] __list_add_debug+0x2e/0xa0
[Fri Nov 10 14:33:45 2017] PGD fd8048066 

[Fri Nov 10 14:33:45 2017] Oops: 0000 [#1] SMP
[Fri Nov 10 14:33:45 2017] Modules linked in: target_core_user target_core_pscsi target_core_file target_core_iblock veth vhost_net vhost macvtap macvlan tun ebtable_filter ebtables ip6table_filter ip6_tables iptable_filter iscsi_target_mod serpent_sse2_x86_64 serpent_generic lrw glue_helper ablk_helper cryptd jc42 nls_iso8859_1 nls_cp437 vfat fat ext4 crc16 jbd2 fscrypto mbcache ttm drm_kms_helper drm gpio_ich psmouse iTCO_wdt iTCO_vendor_support ppdev input_leds i2c_algo_bit intel_powerclamp uio_pdrv_genirq uio fb_sys_fops syscopyarea coretemp pcspkr evdev intel_cstate led_class joydev mousedev sysfillrect mac_hid sysimgblt usbkbd usbmouse i5500_temp i7core_edac ioatdma i2c_i801 i2c_smbus lpc_ich dca parport_pc shpchp edac_core parport button ipmi_ssif acpi_cpufreq sch_fq_codel 8021q garp mrp iscsi_scst scst_local
[Fri Nov 10 14:33:45 2017]  scst dlm dm_thin_pool dm_persistent_data dm_bio_prison dm_bufio libcrc32c crc32c_generic kvm_intel kvm irqbypass ipmi_devintf ipmi_si ipmi_msghandler br_netfilter bridge stp llc ip_tables x_tables zfs(POE) zunicode(POE) zcommon(POE) znvpair(POE) zavl(POE) icp(POE) spl(OE) zlib_deflate xts gf128mul algif_skcipher af_alg dm_crypt dm_mod sd_mod mlx4_en hid_generic usbhid hid uas usb_storage uhci_hcd serio_raw atkbd libps2 crc32c_intel mlx4_core ahci devlink libahci ehci_pci ehci_hcd mpt3sas usbcore e1000e usb_common raid_class ptp scsi_transport_sas pps_core i8042 serio fjes
[Fri Nov 10 14:33:45 2017] CPU: 3 PID: 25439 Comm: worker Tainted: P           OE   4.9.61 #1
[Fri Nov 10 14:33:45 2017] Hardware name: Intel Thurley/Greencity, BIOS GEMAV200 01/28/2011
[Fri Nov 10 14:33:45 2017] task: ffff880f530ad800 task.stack: ffffc9002e1e0000
[Fri Nov 10 14:33:45 2017] RIP: 0010:[<ffffffff813d2cae>]  [<ffffffff813d2cae>] __list_add_debug+0x2e/0xa0
[Fri Nov 10 14:33:45 2017] RSP: 0018:ffffc9002e1e3a20  EFLAGS: 00010246
[Fri Nov 10 14:33:45 2017] RAX: 0000000000000000 RBX: ffff880808d1b138 RCX: 0000000000000000
[Fri Nov 10 14:33:45 2017] RDX: ffff880808d1b138 RSI: 0000000000000000 RDI: ffffc9002e1e3a70
[Fri Nov 10 14:33:45 2017] RBP: ffffc9002e1e3a38 R08: d2b5d6dc2f22ceb0 R09: ffff880e33821800
[Fri Nov 10 14:33:45 2017] R10: ffff880790826a00 R11: 0000000000000000 R12: ffffc9002e1e3a70
[Fri Nov 10 14:33:45 2017] R13: 0000000000000000 R14: 00000000ffffffff R15: ffff880808d1b138
[Fri Nov 10 14:33:45 2017] FS:  000003991f1c3700(0000) GS:ffff88081fac0000(0000) knlGS:0000000000000000
[Fri Nov 10 14:33:45 2017] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[Fri Nov 10 14:33:45 2017] CR2: 0000000000000000 CR3: 0000000001d60000 CR4: 00000000000026f0
[Fri Nov 10 14:33:45 2017] Stack:
[Fri Nov 10 14:33:45 2017]  ffffc9002e1e3a70 0000000000000000 ffff880808d1b138 ffffc9002e1e3a60
[Fri Nov 10 14:33:46 2017]  ffffffff813d2d53 ffff880808d1b130 ffff880f530ad800 ffff880808d1b134
[Fri Nov 10 14:33:46 2017]  ffffc9002e1e3ab8 ffffffff81809131 0000000002404200 0000000000000000
[Fri Nov 10 14:33:46 2017] Call Trace:
[Fri Nov 10 14:33:46 2017]  [<ffffffff813d2d53>] __list_add+0x33/0x60
[Fri Nov 10 14:33:46 2017]  [<ffffffff81809131>] __mutex_lock_slowpath+0xc1/0x130
[Fri Nov 10 14:33:46 2017]  [<ffffffff818091cd>] mutex_lock+0x2d/0x50
[Fri Nov 10 14:33:46 2017]  [<ffffffffa0633747>] zfs_range_lock+0x87/0x5a0 [zfs]
[Fri Nov 10 14:33:46 2017]  [<ffffffffa06623b2>] zvol_get_data+0x82/0x180 [zfs]
[Fri Nov 10 14:33:46 2017]  [<ffffffffa064c9a7>] zil_commit+0xa57/0xd00 [zfs]
[Fri Nov 10 14:33:46 2017]  [<ffffffffa06645d9>] zvol_request+0x189/0x350 [zfs]
[Fri Nov 10 14:33:46 2017]  [<ffffffff8134df73>] generic_make_request+0x133/0x300
[Fri Nov 10 14:33:46 2017]  [<ffffffff8134e18e>] submit_bio+0x4e/0x160
[Fri Nov 10 14:33:46 2017]  [<ffffffff81343bb7>] submit_bio_wait+0x67/0xa0
[Fri Nov 10 14:33:46 2017]  [<ffffffff81352dd5>] blkdev_issue_flush+0x75/0xb0
[Fri Nov 10 14:33:46 2017]  [<ffffffff81249c75>] blkdev_fsync+0x55/0x70
[Fri Nov 10 14:33:46 2017]  [<ffffffff81240979>] vfs_fsync_range+0x59/0xc0
[Fri Nov 10 14:33:46 2017]  [<ffffffff81240a7b>] do_fsync+0x4b/0x80
[Fri Nov 10 14:33:46 2017]  [<ffffffff81240e16>] sys_fdatasync+0x26/0x40
[Fri Nov 10 14:33:46 2017]  [<ffffffff8100191f>] do_syscall_64+0x7f/0x1b0
[Fri Nov 10 14:33:46 2017]  [<ffffffff8180c425>] entry_SYSCALL64_slow_path+0x25/0x25
[Fri Nov 10 14:33:46 2017] Code: e5 41 55 41 54 53 48 81 ec 20 10 00 00 48 83 0c 24 00 48 81 c4 20 10 00 00 48 89 d3 49 89 fc 49 89 f5 48 8b 43 08 4c 39 e8 75 24 <48> 8b 10 48 39 da 75 33 4c 39 e0 74 42 4c 39 e2 74 3d 5b b8 01 
[Fri Nov 10 14:33:46 2017] RIP  [<ffffffff813d2cae>] __list_add_debug+0x2e/0xa0
[Fri Nov 10 14:33:46 2017]  RSP <ffffc9002e1e3a20>
[Fri Nov 10 14:33:46 2017] CR2: 0000000000000000
[Fri Nov 10 14:33:46 2017] ---[ end trace 97dcc88aa692fc35 ]---
[Fri Nov 10 14:33:46 2017] grsec: banning user with uid 99 until system restart for suspicious kernel crash
```
It seems to happen almost randomly within a 24h span. I think one of the underlying SSDs in a RAIDZ pool might be lagging (noticed its UUID in systemd's boot messages indicting slow "spin-up"), but a bad drive in a pool shouldn't merc a system...

sempervictus commented:
@prakashsurya @behlendorf: those DVA errors have something to do with master. There might be a problem after the C99 cleanup...
I rolled back to 20171020 and pulled in the crypto fixes and dbuf_impl stuff. That was a longer running test stack, containing the dmu prefetch, lock tracking removal in SPL and ZFS, this PR, and some other odds and ends + pulls from master. I've included the commit difference between this working branch and master in case something jumps out as possibly affecting the DVA problem:
```
git log --graph --pretty=format:'%Cred%h%Creset -%C(yellow)%d%Creset %s %Cgreen(%cr)%Creset' --abbrev-commit --date=relative HEAD..origin/master
* 62df1bc81 - (origin/master, origin/HEAD, master) Fix encryption root hierarchy issue (2 days ago)
* 71a24c3c5 - Handle compressed buffers in __dbuf_hold_impl() (2 days ago)
* eef005d88 - Refresh TEST file to include new variables (2 days ago)
* 80b485246 - Cleanup systemd dependencies (2 days ago)
* 011ef12c7 - Fix undefined %{systemd_svcs} in RPM scriptlets (2 days ago)
* d8fdfc2d6 - OpenZFS 8607 - variable set but not used (2 days ago)
* 87fbf4363 - Provide tags in perf-regression.run (3 days ago)
* 271955da3 - Fix zfs-tests.sh single test functionality (3 days ago)
* cb3b0419b - contrib/initramfs: switch to automake (3 days ago)
* a3df7fa79 - Bug fix in qat_compress.c when compressed size is < 4KB (3 days ago)
* 681957fe2 - Sort output of tunables in arc_summary.py (3 days ago)
* 12954494a - Disable automatic dependencies in DKMS package (3 days ago)
* 31df97cda - Build regression in c89 cleanups (3 days ago)
* b58b73ce7 - Disable zpool_import_missing_003_pos (3 days ago)
* 23ea00a1f - (datto/master) Add documentation strings to arc_summary.py (5 days ago)
* 2df9ad1c0 - Fix column alignment with long zpool names (5 days ago)
* cd1813d36 - Rewrite fHits() in arc_summary.py with SI units (6 days ago)
* 1c27024e2 - Undo c89 workarounds to match with upstream (6 days ago)
* df1f129bc - Minor code cleanup in arc_summary.py (7 days ago)
*   34c2b3680 - Initramfs fixes (7 days ago)
|\  
| * 4fc411f7a - (origin/pr/6807) initramfs: Set elevator=noop on the rpool's disks (9 days ago)
| * 11b9dcfb2 - initramfs: Fix a spelling error (9 days ago)
| * 4767c7a14 - initramfs: Fix inconsistent whitespace (9 days ago)
* | 9a810efb0 - Allow test-runner to filter test groups by tag (7 days ago)
|/  
* c9427c469 - Add scan.coverity.com badge to README (11 days ago)
* f3c8c9e6f - OpenZFS 640 - number_to_scaled_string is duplicated in several commands (11 days ago)
* 47c8e7fd9 - Rewrite of function fBytes() in arc_summary.py (11 days ago)
* 5c2552c56 - systemd zfs-import.target and documentation (11 days ago)
* ca85d6909 - Update zfs module parameters man5 (11 days ago)
* f4ae39a19 - Fix status command options in zpool(8) (2 weeks ago)
* 867959b58 - OpenZFS 8081 - Compiler warnings in zdb (2 weeks ago)
* a94d38c0f - Correct make mancheck recipe (2 weeks ago)
* ee45fbd89 - ZFS send fails to dump objects larger than 128PiB (2 weeks ago)
* 88f9c9396 - Allow 'zpool events' filtering by pool name (2 weeks ago)
* a032ac4b3 - OpenZFS 8558, 8602 - lwp_create() returns EAGAIN (2 weeks ago)
* d3f2cd7e3 - Added no_scrub_restart flag to zpool reopen (2 weeks ago)
* 3ad59c015 - arcstat: flush stdout / outfile after each line (2 weeks ago)
* 69b229bd6 - commitcheck: Multiple OpenZFS ports in commit (2 weeks ago)
* 8dcaf243d - Add Coverity defect fix commit checker support (2 weeks ago)
* 64b8c58e3 - Ensure arc_size_break is filled in arc_summary.py (3 weeks ago)
* 63e5e960b - Correct flake8 errors after STYLE builder update (3 weeks ago)
* d9daa7abc - ZTS: Add auto-spare tests (3 weeks ago)
* f8cd871a0 - Use ashift=12 by default on SSDSC2BW48 disks (3 weeks ago)
* 70c8a7944 - Provide commit message format for Coverity defects (3 weeks ago)
* d5e024cba - Emit history events for 'zpool create' (3 weeks ago)
```
This PR is included in the working branch, so i dont think its the culprit for that issue. I'll have to wait and see if the hardware failure condition comes  up in the next day on the second one - sort of hard to repro.

sempervictus commented:
"All together now..." The hardware system did crash on the new test stack again, this time displaying both the DVA issue and the zil_commit problem:
```
[Sat Nov 11 01:07:15 2017] PANIC: metaslab_free_dva(): bad DVA 1573890:16998400:1587590144
[Sat Nov 11 01:07:15 2017] Showing stack for process 5182
[Sat Nov 11 01:07:15 2017] CPU: 1 PID: 5182 Comm: dp_sync_taskq Tainted: P           OE   4.9.61 #1
[Sat Nov 11 01:07:15 2017] Hardware name: Intel Thurley/Greencity, BIOS GEMAV200 01/28/2011
[Sat Nov 11 01:07:15 2017]  ffffc9003bc0f820 ffffffff813a6609 0000000000000003 ffffffffa0704238
[Sat Nov 11 01:07:15 2017]  ffffc9003bc0f830 ffffffffa00feb63 ffffc9003bc0f968 ffffffffa00fed57
[Sat Nov 11 01:07:15 2017]  ffff880fe4ede380 62616c736174656d 76645f656572665f 646162203a292861
[Sat Nov 11 01:07:15 2017] Call Trace:
[Sat Nov 11 01:07:15 2017]  [<ffffffff813a6609>] dump_stack+0x68/0x8f
[Sat Nov 11 01:07:15 2017]  [<ffffffffa0704238>] ? _fini+0x24b99/0x6a98d [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa00feb63>] spl_dumpstack+0x43/0x50 [spl]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa00fed57>] vcmn_err+0xd7/0x110 [spl]
[Sat Nov 11 01:07:15 2017]  [<ffffffff810af741>] ? enqueue_task_fair+0xc1/0x990
[Sat Nov 11 01:07:15 2017]  [<ffffffff810a705c>] ? sched_clock_cpu+0xac/0xd0
[Sat Nov 11 01:07:15 2017]  [<ffffffff8109f069>] ? check_preempt_curr+0xa9/0xc0
[Sat Nov 11 01:07:15 2017]  [<ffffffff811d6c77>] ? kmem_cache_alloc+0x107/0x160
[Sat Nov 11 01:07:15 2017]  [<ffffffffa00f9f36>] ? spl_kmem_cache_alloc+0x76/0x720 [spl]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa0704238>] ? _fini+0x24b99/0x6a98d [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa06051db>] zfs_panic_recover+0x6b/0x90 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffff818091c0>] ? mutex_lock+0x20/0x50
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05e3e4b>] metaslab_free_dva+0x17b/0x410 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05e7e40>] metaslab_free+0xf0/0x130 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa065f1a4>] zio_dva_free+0x24/0x40 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa066417a>] zio_wait+0xba/0x1d0 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa066432b>] zio_free+0x9b/0xf0 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05dd280>] dsl_free+0x20/0x30 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05bc89b>] dsl_dataset_block_kill+0x27b/0x480 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05baf66>] free_blocks+0xf6/0x1e0 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05bb380>] ? free_children+0x330/0x330 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05bb5ae>] dnode_sync_free_range+0x22e/0x290 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa0647c79>] ? zfs_space_delta_cb+0xa9/0x1d0 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05bb380>] ? free_children+0x330/0x330 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05eb2aa>] range_tree_vacate+0x7a/0xe0 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05bbc51>] dnode_sync+0x2c1/0x8e0 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffff813d2d53>] ? __list_add+0x33/0x60
[Sat Nov 11 01:07:15 2017]  [<ffffffffa05a5a67>] sync_dnodes_task+0x77/0xb0 [zfs]
[Sat Nov 11 01:07:15 2017]  [<ffffffffa00fcca1>] taskq_thread+0x2b1/0x4d0 [spl]
[Sat Nov 11 01:07:15 2017]  [<ffffffff810a0160>] ? wake_up_q+0x90/0x90
[Sat Nov 11 01:07:15 2017]  [<ffffffffa00fc9f0>] ? task_done+0xa0/0xa0 [spl]
[Sat Nov 11 01:07:15 2017]  [<ffffffff81095ddd>] kthread+0xfd/0x120
[Sat Nov 11 01:07:15 2017]  [<ffffffff81095ce0>] ? kthread_parkme+0x40/0x40
[Sat Nov 11 01:07:15 2017]  [<ffffffff8180c5e7>] ret_from_fork+0x37/0x50
[Sat Nov 11 01:11:00 2017] general protection fault: 0000 [#1] SMP
[Sat Nov 11 01:11:01 2017] Modules linked in: target_core_user target_core_pscsi target_core_file target_core_iblock veth vhost_net vhost macvtap macvlan tun ebtable_filter ebtables ip6table_filter ip6_tables iptable_filter iscsi_target_mod serpent_sse2_x86_64 serpent_generic lrw glue_helper ablk_helper cryptd jc42 nls_iso8859_1 nls_cp437 vfat fat ext4 crc16 jbd2 fscrypto mbcache iTCO_wdt intel_powerclamp iTCO_vendor_support gpio_ich psmouse ttm ppdev drm_kms_helper coretemp uio_pdrv_genirq pcspkr drm intel_cstate uio joydev i2c_algo_bit evdev input_leds mousedev fb_sys_fops syscopyarea sysfillrect sysimgblt led_class mac_hid usbmouse usbkbd i2c_i801 i7core_edac edac_core ioatdma i2c_smbus i5500_temp shpchp lpc_ich parport_pc parport button dca acpi_cpufreq ipmi_ssif sch_fq_codel 8021q garp mrp iscsi_scst scst_local
[Sat Nov 11 01:11:01 2017]  scst dlm dm_thin_pool dm_persistent_data dm_bio_prison dm_bufio libcrc32c crc32c_generic kvm_intel kvm irqbypass ipmi_devintf ipmi_si ipmi_msghandler br_netfilter bridge stp llc ip_tables x_tables zfs(POE) zunicode(POE) zcommon(POE) znvpair(POE) zavl(POE) icp(POE) spl(OE) zlib_deflate xts gf128mul algif_skcipher af_alg dm_crypt dm_mod sd_mod mlx4_en hid_generic usbhid hid uas usb_storage uhci_hcd serio_raw atkbd libps2 crc32c_intel mlx4_core devlink ahci ehci_pci libahci ehci_hcd usbcore mpt3sas e1000e usb_common raid_class ptp scsi_transport_sas pps_core i8042 serio fjes
[Sat Nov 11 01:11:01 2017] CPU: 14 PID: 3509 Comm: mysqld Tainted: P           OE   4.9.61 #1
[Sat Nov 11 01:11:01 2017] Hardware name: Intel Thurley/Greencity, BIOS GEMAV200 01/28/2011
[Sat Nov 11 01:11:01 2017] task: ffff880fd6ef8000 task.stack: ffffc9000c6a4000
[Sat Nov 11 01:11:01 2017] RIP: 0010:[<ffffffffa065329b>]  [<ffffffffa065329b>] zfs_get_data+0x3b/0x2a0 [zfs]
[Sat Nov 11 01:11:01 2017] RSP: 0018:ffffc9000c6a7c18  EFLAGS: 00010246
[Sat Nov 11 01:11:01 2017] RAX: 0000000000000000 RBX: ffff880b33050178 RCX: ffff880b80b7b780
[Sat Nov 11 01:11:01 2017] RDX: ffffc9000c6a7c38 RSI: 000000000000202f RDI: fefefefefefefefe
[Sat Nov 11 01:11:01 2017] RBP: ffffc9000c6a7c78 R08: ffff880b0d326220 R09: ffff880b330500b8
[Sat Nov 11 01:11:01 2017] R10: 0000000000000000 R11: 0000000000000000 R12: fefefefefefefefe
[Sat Nov 11 01:11:01 2017] R13: ffff880b330500b8 R14: ffff880b80b7b780 R15: ffff880b80b7b780
[Sat Nov 11 01:11:02 2017] FS:  00000338b65ba700(0000) GS:ffff88101f380000(0000) knlGS:0000000000000000
[Sat Nov 11 01:11:02 2017] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[Sat Nov 11 01:11:02 2017] CR2: 00007f6016d36228 CR3: 0000000001d76000 CR4: 00000000000026f0
[Sat Nov 11 01:11:02 2017] Stack:
[Sat Nov 11 01:11:02 2017]  ffff880b0d326220 ffff880568deef78 0000000000000b1e 0000000000000000
[Sat Nov 11 01:11:02 2017]  fffffffffffffffe 000000000001829a b3b9eca414349ea6 0000000000001000
[Sat Nov 11 01:11:02 2017]  0000000000001000 ffff880811bd4000 ffff880b80b7b780 00000000000000c0
[Sat Nov 11 01:11:02 2017] Call Trace:
[Sat Nov 11 01:11:02 2017]  [<ffffffffa065de60>] zil_commit+0xb00/0xdd0 [zfs]
[Sat Nov 11 01:11:02 2017]  [<ffffffffa06508b3>] zfs_fsync+0x73/0xe0 [zfs]
[Sat Nov 11 01:11:02 2017]  [<ffffffffa066efb4>] zpl_fsync+0x74/0xb0 [zfs]
[Sat Nov 11 01:11:02 2017]  [<ffffffff81240979>] vfs_fsync_range+0x59/0xc0
[Sat Nov 11 01:11:02 2017]  [<ffffffff81240a7b>] do_fsync+0x4b/0x80
[Sat Nov 11 01:11:02 2017]  [<ffffffff81240de3>] sys_fsync+0x23/0x30
[Sat Nov 11 01:11:02 2017]  [<ffffffff8100191f>] do_syscall_64+0x7f/0x1b0
[Sat Nov 11 01:11:02 2017]  [<ffffffff8180c425>] entry_SYSCALL64_slow_path+0x25/0x25
[Sat Nov 11 01:11:02 2017] Code: 41 54 49 89 fc 53 48 89 d3 48 83 ec 38 4c 89 45 a0 65 48 8b 04 25 28 00 00 00 48 89 45 d0 31 c0 49 8b 75 20 48 8d 55 c0 4c 89 e7 <49> 8b 44 24 18 4d 8b 75 28 48 89 75 a8 48 89 45 b0 49 8b 45 30 
[Sat Nov 11 01:11:02 2017] RIP  [<ffffffffa065329b>] zfs_get_data+0x3b/0x2a0 [zfs]
[Sat Nov 11 01:11:02 2017]  RSP <ffffc9000c6a7c18>
[Sat Nov 11 01:11:02 2017] ---[ end trace 340bd621992fcf74 ]---
[Sat Nov 11 01:11:02 2017] grsec: banning user with uid 498 until system restart for suspicious kernel crash
```
Based on the time stampts, it looks like the DVA problem hit, then the zil_commit thing happened, possibily as a result of the DVA problem. Any hints on what i should be looking at in that context?
Im rebooting that specific host which seems to exhibit all the things on a master-only build of SPL and ZFS, and letting it shake out another 24h before trying this again. Unfortunately this _is_ a production system housing one 3rd of a pacemaker cluster, so it cant stay in test-mode perpetually.

sempervictus commented:
Does this PR change on-disk structures at all in the ZIL? Trying to import a pool which suffered a crash using master-only builds results in:
```
[Sun Nov 12 01:51:55 2017] NMI watchdog: BUG: soft lockup - CPU#0 stuck for 22s! [dmu_objset_find:6907]
[Sun Nov 12 01:51:55 2017] Modules linked in: serpent_sse2_x86_64 serpent_generic lrw glue_helper ablk_helper cryptd jc42 nls_iso8859_1 nls_cp437 vfat fat ext4 crc16 jbd2 fscrypto mbcache iTCO_wdt iTCO_vendor_support gpio_ich ppdev ttm drm_kms_helper intel_powerclamp drm coretemp intel_cstate input_leds evdev i2c_algo_bit fb_sys_fops led_class joydev syscopyarea mousedev mac_hid sysfillrect usbmouse usbkbd sysimgblt psmouse uio_pdrv_genirq pcspkr uio i2c_i801 parport_pc i2c_smbus ioatdma i7core_edac parport button edac_core shpchp i5500_temp ipmi_ssif dca lpc_ich acpi_cpufreq sch_fq_codel 8021q garp mrp iscsi_scst scst_local scst dlm dm_thin_pool dm_persistent_data dm_bio_prison dm_bufio libcrc32c crc32c_generic kvm_intel kvm irqbypass ipmi_devintf ipmi_si ipmi_msghandler br_netfilter bridge stp llc ip_tables x_tables
[Sun Nov 12 01:51:55 2017]  zfs(POE) zunicode(POE) zcommon(POE) znvpair(POE) zavl(POE) icp(POE) spl(OE) zlib_deflate xts gf128mul algif_skcipher af_alg dm_crypt dm_mod sd_mod mlx4_en hid_generic usbhid hid uas usb_storage uhci_hcd serio_raw atkbd libps2 crc32c_intel mlx4_core ahci devlink libahci ehci_pci ehci_hcd usbcore e1000e mpt3sas usb_common ptp raid_class scsi_transport_sas pps_core i8042 serio fjes
[Sun Nov 12 01:51:55 2017] CPU: 0 PID: 6907 Comm: dmu_objset_find Tainted: P           OEL  4.9.61 #1
[Sun Nov 12 01:51:55 2017] Hardware name: Intel Thurley/Greencity, BIOS GEMAV200 01/28/2011
[Sun Nov 12 01:51:55 2017] task: ffff8807eb65ee00 task.stack: ffffc9003e910000
[Sun Nov 12 01:51:55 2017] RIP: 0010:[<ffffffffa0626649>]  [<ffffffffa0626649>] zil_claim_log_record+0x19/0x60 [zfs]
[Sun Nov 12 01:51:55 2017] RSP: 0000:ffffc9003e913a38  EFLAGS: 00000246
[Sun Nov 12 01:51:55 2017] RAX: 0000000000000000 RBX: ffffc90036d86efe RCX: 0000000002305788
[Sun Nov 12 01:51:55 2017] RDX: 0000000000000000 RSI: ffffc90036d86efe RDI: ffff880813d56800
[Sun Nov 12 01:51:55 2017] RBP: ffffc9003e913a58 R08: ffff8807e0324b58 R09: 00000001802e001c
[Sun Nov 12 01:51:55 2017] R10: 0100000000000000 R11: 0100000000000000 R12: ffff880813d56800
[Sun Nov 12 01:51:55 2017] R13: 0000000000000000 R14: 0000000002305788 R15: ffffffffffffffff
[Sun Nov 12 01:51:55 2017] FS:  0000000000000000(0000) GS:ffff88081fa00000(0000) knlGS:0000000000000000
[Sun Nov 12 01:51:55 2017] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[Sun Nov 12 01:51:55 2017] CR2: 00000061b0b41000 CR3: 0000000001d5a000 CR4: 00000000000006f0
[Sun Nov 12 01:51:55 2017] Stack:
[Sun Nov 12 01:51:55 2017]  0000000000000000 ffffffffa0626630 ffffc90036d86efe 00000002540ee182
[Sun Nov 12 01:51:55 2017]  ffffc9003e913c68 ffffffffa06280b8 ffffc90001800190 ffffc9003e913aec
[Sun Nov 12 01:51:55 2017]  ffffc9003e913af8 ffffffffa0626630 ffffc90037d97000 0000000000000000
[Sun Nov 12 01:51:55 2017] Call Trace:
[Sun Nov 12 01:51:55 2017]  [<ffffffffa0626630>] ? zil_claim_log_block+0x90/0x90 [zfs]
[Sun Nov 12 01:51:55 2017]  [<ffffffffa06280b8>] zil_parse+0x688/0x910 [zfs]
[Sun Nov 12 01:51:55 2017]  [<ffffffffa0626630>] ? zil_claim_log_block+0x90/0x90 [zfs]
[Sun Nov 12 01:51:55 2017]  [<ffffffffa06265a0>] ? zil_replaying+0x70/0x70 [zfs]
[Sun Nov 12 01:51:55 2017]  [<ffffffffa0628d91>] zil_check_log_chain+0xe1/0x1a0 [zfs]
[Sun Nov 12 01:51:55 2017]  [<ffffffffa0574e11>] dmu_objset_find_dp_impl+0x151/0x400 [zfs]
[Sun Nov 12 01:51:55 2017]  [<ffffffffa05750e6>] dmu_objset_find_dp_cb+0x26/0x40 [zfs]
[Sun Nov 12 01:51:55 2017]  [<ffffffffa014cca1>] taskq_thread+0x2b1/0x4d0 [spl]
[Sun Nov 12 01:51:55 2017]  [<ffffffff810a0160>] ? wake_up_q+0x90/0x90
[Sun Nov 12 01:51:55 2017]  [<ffffffffa014c9f0>] ? task_done+0xa0/0xa0 [spl]
[Sun Nov 12 01:51:55 2017]  [<ffffffff81095ddd>] kthread+0xfd/0x120
[Sun Nov 12 01:51:55 2017]  [<ffffffff81095ce0>] ? kthread_parkme+0x40/0x40
[Sun Nov 12 01:51:55 2017]  [<ffffffff8180c5e7>] ret_from_fork+0x37/0x50
```

sempervictus commented:
Good news: while my pool is dead as a doornail, the cause did not come from here. I rebuilt and retested with all the PRs/commits in my test stack atop master and seem to have localized this to the datto work around scrubs and the likely abandoned crypt fmt fix branch. I'll keep a long running zloop on this build to try and work out any kinks, but far as my issues go, i'm 99% sure this didnt cause them, just happened to be in play when the fan was hit.

behlendorf commented:
@sempervictus no ZIL format changes in this PR or the scrub work.  If you do have any suspicions about the current scrub patches please let me know now since those will likely be merged this week.

behlendorf commented:
@prakashsurya I was looking at the `ztest` ASSERT from [CentOS 6](http://build.zfsonlinux.org/builders/CentOS%206%20x86_64%20%28TEST%29/builds/933/steps/shell_5/logs/ztest.out) in `zil_lwb_write_issue`.  Specifically it ASSERT'ed because `dmu_tx_assign` returned `ERESTART` due to a suspended pool.  While far from likely there are a few other ways `ERESTART` can be returned here.

In all of these rare cases it looks like the right time to do is abort the TX and return NULL.  What do you think about removing this ASSERT and updating the comment accordingly.  Incidentally this isn't a new issue introduced by this PR, but since we've been focusing on the ZIL we may be just more aware of it.

```
error == 5 (0x55 == 0x5)
ASSERT at ../../module/zfs/zil.c:1304:zil_lwb_write_issue()/sbin/ztest[0x407c13]
/lib64/libpthread.so.0(+0xf7e0)[0x7f8f078707e0]
/lib64/libc.so.6(gsignal+0x35)[0x7f8f074ff495]
/lib64/libc.so.6(abort+0x175)[0x7f8f07500c75]
/lib64/libzpool.so.2(+0x162c33)[0x7f8f0868ac33]
/lib64/libzpool.so.2(+0x1669b8)[0x7f8f0868e9b8]
/lib64/libzpool.so.2(zil_commit+0x9fe)[0x7f8f0869119e]
/sbin/ztest[0x408a87]
/sbin/ztest[0x408ddd]
/lib64/libpthread.so.0(+0x7aa1)[0x7f8f07868aa1]
/lib64/libc.so.6(clone+0x6d)[0x7f8f075b5bcd]
child died with signal 6
```

behlendorf commented:
@sempervictus I believe @prakashsurya's fix in e93b150 could explain you corrupt ZIL record.

sempervictus commented:
Could that result in the bad dva issue?

behlendorf commented:
@sempervictus I believe so under an unlikely set of circumstances.  You'd need to write a ZIL record which included a dva which was damaged due to the memory it was stored in being free'd and reused as described in e93b150.  Then you'd have to replay that damaged ZIL record on mount.  You'd need to be really really unlucky, but I think it could happen.

prakashsurya commented:
Now that I have a working implementation, I was able to run the same performance tests that I used on illumos, to test this PR on Ubuntu 16.04. See [here](https://www.prakashsurya.com/post/2017-11-17-performance-testing-results-for-zfsonlinux-6566/) for the results.

@behlendorf we can discuss the results in detail after the holiday, but the SSD numbers are a little concerning. I've copied the graphs of the percentage change in latency between master and this PR below (these are also contain in the jupyter notebooks which I linked to above):

### Using HDDs -- each `fio` thread submitting writes as fast as possible:
![max-rate-submit-hdd-latency](https://user-images.githubusercontent.com/658081/32972063-2409c8b4-cba5-11e7-83ba-ac5f2d92394e.png)

### Using HDDs -- each `fio` thread submitting 64 writes per second:
![fixed-rate-submit-hdd-latency](https://user-images.githubusercontent.com/658081/32972065-2900695e-cba5-11e7-8af8-47285103ef82.png)

### Using SSDs -- each `fio` thread submitting writes as fast as possible:
![max-rate-submit-ssd-latency](https://user-images.githubusercontent.com/658081/32971947-9a8d8d96-cba4-11e7-9c7b-90df27f10b44.png)

### Using SSDs -- each `fio` thread submitting 64 writes per second:
![fixed-rate-submit-ssd-latency](https://user-images.githubusercontent.com/658081/32971969-aa664ffa-cba4-11e7-8e93-1f06470203d3.png)

sempervictus commented:
In the graphs, does the term "using" indicate for slog or for data?

sempervictus commented:
@prakashsurya: the pool which keeps getting "murdolated" in #6856 is a 5 member RAIDZ1 pool consisting of 250G Samsung EVO 850s on a SAS HBA. A 5-wide 2-way mirror span running its disks and cache from the same HBA (SLOG are m2's mirrored on PCI for that pool), is unaffected during these events, as is a 4 member RAID1 1T EVO 8 50 pool (but that's seeing no IO, and is on separate datapaths internally). Its getting killed, repeatedly, on a build consisting of nothing but master (at time of build) and this PR. I just refreshed the system with the same thing built a few hours ago with your updates along with those that went into master, destroyed the unimportable pool (again), rebuilt it, reseeded it with the VMs and containers it needed from their last known good snap, and lost the thing again in a matter of days. Something's rotten in the state of Denmark, and i'm apparently down-wind.
I cant say for sure that this PR is causing that failure, but whatever IS going on, its requiring manual recovery efforts in a place about as automated as Megatron's bowels, at a somewhat less -than-optimal time :-). Be glad to keep going through this till Monday, but i'll have to get a master branch build at that point if it keeps happening and we dont know why (identical system is still running an older revision of this PR, as are many others, with none of the stack traces related to this unfortunate set of events).

prakashsurya commented:
@sempervictus For my tests, I was *not* using an SLOG. "Using" simply means the disks used by the zpool itself; i.e. I had pools with 1 to 8 HDDs, and pools with 1 to 8 SSDs (no redundancy, no SLOG).

As far as your failures.. you're still seeing problems with the latest code of this PR, with my follow-up fixes? Are the issues seen on ZIL replay (i.e. initial mount), or after the FS/dataset is mounted?

sempervictus commented:
@prakashsurya: i've had to pull this PR out of the stack because it seems to be causing two distinct issues:
1 - ARC, for some reason, starts regulating itself despite sysctl settings. i have no idea how in the world this is happening from this PR, but dropping it from the stack restores normal behavior.
2 - With the rebased code, a RAIDZ1 pool consisting entirely of SSDs went bad on me once. Its running with this PR, currently, and sending its deltas hourly to another system (hoping to "validate in flight"). I am thinking that the issues, both times, were caused by IO from ZVOLs backing control nodes for openstack (MySQL, Rabbit, etc, in KVM instances on raw ZVOLs). That particular pool also serves an LXC container with Zenoss doing all the same MySQL and Rabbit stuff, along with a ton of file writes via the ZPL as opposed to ZVOL path. Errors always seem to start from ZVOL ops though, so the ZPL IO may simply be adding contenttion. Same pool went bad before the rebase as well, so i'm not sure if the cause of the error was what you fixed, or what i'm now thinking #6294 is designed to fix (#6888). Reason for this logic is that what i saw last night didnt have this PR in it, on a wholly separate system, in a another location. It was just a master build, and too much IO to a ZVOL killed in that specific manner. I do however suspect that the write improvements here are causing those failures to amplify as opposed to the "serialized failure" without the optimizations here, potentially explaining my bad pools. Do scrubs find DVA errors?

behlendorf commented:
@sempervictus were you able to confirm that removing this PR from your stack resolved the two issue you mentioned?

> Do scrubs find DVA errors?

It depends exactly what;'s wrong with the DVA but in general yes.

dweeezil commented:
I've been loosely following this issue, along with #6874, and it seems that the bad blkptrs are being written to the pool on systems which _do_ have ECC memory.  In that case, it might be useful to put some calls to ```zfs_blkptr_verify()``` in the _write_ path,  potentially even in ```zio_create()``` (when ```bp``` isn't null) to try to catch the condition under which they're being written.

behlendorf commented:
@dweeezil good thought.  For debugging it might be useful to additionally verify the blkptr in `zio_ready` and `zio_done`.

prakashsurya commented:
@behlendorf I rebased this onto the latest master, and squashed to a single commit. The only failure I see is the following, from the 32-bit tester:
```
Test: /usr/share/zfs/zfs-tests/tests/functional/vdev_zaps/vdev_zaps_004_pos (run as root) [00:03] [FAIL]
23:16:12.42 ASSERTION: Per-vdev ZAPs are transferred properly on attach/detach
23:16:12.56 SUCCESS: zpool create -f testpool loop0
23:16:15.42 SUCCESS: zpool attach testpool loop0 loop1
23:16:15.81 zdb: can't open 'testpool': No such device or address
```
Is this a known flakey test? This doesn't seem related to my changes at first glance.

behlendorf commented:
@prakashsurya thanks!  That's a known spurious failure, I've resubmitting that build.
